### DISCUSSION TITLE
How can I use a custom inkscape location?

### DISCUSSION THREAD

**Discussion by BeatVids** [ImageMagick/ImageMagick#7116] [Help] [answered]
### ImageMagick version

7.1.1-28 Q16-HDRI aarch64 21967

### Operating system, version and so on

Android phone with Termux environment (non-rooted)

### Description

Hi, I'm running ImageMagick via command line on my Android phone using the app Termux. It's pretty neat, and I get to do everything I want so far!

However, I need to configure the location of inkscape, I am getting this error:
`magick: FailedToExecuteCommand 'inkscape'`

My inkscape binary is at `/data/data/com.termux/files/usr/bin/inkscape`

How can I tell magick which specific inkscape location I want to use? I am not rooted, so I'm limited on some things I can do, but hopefully there is a simple solution that I haven't been able to find with a search. Thank you

**Comment by fmw42** [comment]
Add the full path to gs in the PDF/PS/EPS rows in the delegates.xml file in the Imagemagick config folder. For example add your full_path_to/ before gs in the command= part of the following row in the file.

`  <delegate decode="ps" encode="pdf" mode="bi" command="&quot;gs&quot; -sstdout=%%stderr -dQUIET -dSAFER -dBATCH -dNOPAUSE -dNOPROMPT -dMaxBitmap=500000000 -dAlignToPixels=0 -dGridFitTT=2 &quot;-sDEVICE=pdfwrite&quot; &quot;-sOutputFile=%o&quot; &quot;-f%i&quot;"/>
`

**Reply by BeatVids** [reply]
Thanks Fred,

That is a great idea, but I triend looking for delegates.xml, and I don't believe I have access to it, since I couldn't find it with a `find`

I tried putting it in `$HOME/.config/ImageMagick`, but the error is still printing it as `inkscape`

```
cat $HOME/.config/ImageMagick
<delegate decode="svg:decode" stealth="True" command="&quot;/data/data/com.termux/files/usr/bin/inkscape&quot; &quot;%s&quot; --export-png=&quot;%s&quot; --export-dpi=&quot;%s&quot; --export-background=&quot;%s&quot; --export-background-opacity=&quot;%s&quot; &gt; &quot;%s&quot; 2&gt;&amp;1"/>
```

**Comment by fmw42** [comment]
cd to your Imagemagick folder (downloaded).
cd config
cat delegates.xml

**Comment by BeatVids** [comment]
I installed it with the package manager, but I believe I found the config file:

```
cat /data/data/com.termux/files/usr/etc/ImageMagick-7/delegates.xml | grep 'svg:decode'
  <delegate decode="svg:decode" stealth="True" command="&quot;/data/data/com.termux/files/usr/bin/inkscape&quot; &quot;%s&quot; --export-filename=&quot;%s&quot; --export-dpi=&quot;%s&quot; --export-background=&quot;%s&quot; --export-background-opacity=&quot;%s&quot; &gt; &quot;%s&quot; 2&gt;&amp;1"/>
```

I believe it's the right file since now at least my error message involves the new path. Progress!

```
magick in.svg out.png
magick: FailedToExecuteCommand `'/data/data/com.termux/files/usr/bin/inkscape' '/data/data/com.termux/files/usr/tmp/magick-BILLZi3g5Rb3PjMWJBN6eHucJW7cCE5n' --export-filename='/data/data/com.termux/files/usr/tmp/magick-w_Gmp0DL5FCfAHisVVoL-pGs9u696YKg.png' --export-dpi='96' --export-background='rgb(100%,100%,100%)' --export-background-opacity='1' > '/data/data/com.termux/files/usr/tmp/magick-24_UkC0lBLWy3fB0qc2XRtinTDwu8um5' 2>&1' (-1) @ error/delegate.c/ExternalDelegateCommand/522.
```

Why would it be failing at this point? I confirmed standalone inkscape with the full path to work. Thanks for getting me to this point Fred

**Reply by fmw42** [reply]
I do not know.  Sorry about sending you to Ghostscript (gs) rather than Inkscape. I do not know what I was thinking at the time.

**Reply by BeatVids** [reply]
No need to apologize, thanks for the help so far Fred, I'll post a resolution on the thread if I find it.

**Reply by fmw42** [reply]
Unfortunately, I do not know much about using Imagemagick on Android.  Usually the config folder is inside your downloaded Imagemagick folder.

**Comment by snibgo** [comment]
Is the path (directory) that contains Inkscape on the system path? If not, try putting it on the path. I don't know how to do that on Android.

**Reply by BeatVids** [reply]
It's even on the same path as magick

```
whereis magick inkscape
magick: /data/data/com.termux/files/usr/bin/magick
inkscape: /data/data/com.termux/files/usr/bin/inkscape
```

Running that inkscape path is producing valid output file:

`/data/data/com.termux/files/usr/bin/inkscape -w 1024 -h 1024 in.svg -o output.png`

With the updated path on the delegates.xml file, I am still getting the error:

`magick: FailedToExecuteCommand `'/data/data/com.termux/files/usr/bin/inkscape' '/data/data/com.termux/files/usr/tmp/magick-p54mvbSqSK6rh8AlPT_n5BuAIRMIYfyL' --export-filename='/data/data/com.termux/files/usr/tmp/magick-bxx_AkGXSZHrCirzyC90IH168pQjV8U1.png' --export-dpi='96' --export-background='rgb(100%,100%,100%)' --export-background-opacity='1' > '/data/data/com.termux/files/usr/tmp/magick-JK_rD4JW2E-IvhViRFY26G7p3Y617UZO' 2>&1' (-1) @ error/delegate.c/ExternalDelegateCommand/522.`

**Comment by fmw42** [comment]
try snibgo's suggestion.  

echo $PATH

now find the $PATH in some environment variable on Android, perhaps in your home directory in the .profile file and add your path to your Inkscape in the $PATH variable

**Reply by BeatVids** [reply]
```
echo $PATH
/data/data/com.termux/files/usr/bin:/data/data/com.termux/files/usr/bin/texlive:/data/data/com.termux/files/usr/bin:/data/data/com.termux/files/home/bin
```

I also tried it with /data/data/com.termux/files/home/bin/inkscape

No .profile here, any other suggestions?

**Comment by snibgo** [comment]
The directory that contains Inkscape, `/data/data/com.termux/files/usr/bin/`, is on your $PATH, so that isn't the problem.

What is your version of Inkscape?

The IM command uses a temporary directory. Check that you can write large files to `/data/data/com.termux/files/usr/tmp/`, and read them.

You have tried running Inkscape from the command line, and that works. Now try a command more like the one used by IM, which includes `--export-filename` and so on.

**Reply by BeatVids** [reply]
That's a great idea. I just ran:

`inkscape in.svg --export-filename='/data/data/com.termux/files/usr/tmp/magick-bxx_AkGXSZHrCirzyC90IH168pQjV8U1.png' --export-dpi='96' --export-background='rgb(100%,100%,100%)' --export-background-opacity='1' > '/data/data/com.termux/files/usr/tmp/magick-JK_rD4JW2E-IvhViRFY26G7p3Y617UZO'`

`/data/data/com.termux/files/usr/tmp/magick-bxx_AkGXSZHrCirzyC90IH168pQjV8U1.png` is successfully created into a png, so the command does work with standalone inkscape

**Reply by BeatVids** [reply]
However `magick in.svg out.png` still gives the same error (of course since nothing has changed)

**Reply by BeatVids** [reply]
Sorry, forgot to mention:
Inkscape 1.3.2 (091e20ef0f, 2023-11-25)

**Comment by snibgo** [comment]
Perhaps IM can't spawn an external program, for some reason, eg security. To test this, create a simple entry in `delegates.xml` that runs an external program. I suggest:
```
  <delegate decode="myls" command="ls" />
```
"ls" is the Unix file lister. Now run:
```
magick rose: r.png

magick myls:r.png NULL:
```
The first just creates a file `r.png`. The second reads it with the external delegate `ls`. The expected output is a listing of files, and then an error because `ls` has not created an image. That is what happens for me, but you might get an error `FailedToExecuteCommand `.

**Reply by BeatVids** [reply]
```
magick rose: r.png
magick myls:r.png NULL:
r.png
magick: unable to open image '/data/data/com.termux/files/usr/tmp/magick-4GXcHbsfahGflSboUph-0SuocxEB2LHR': No such file or directory @ error/blob.c/OpenBlob/3571.
magick: unable to open file '/data/data/com.termux/files/usr/tmp/magick-4GXcHbsfahGflSboUph-0SuocxEB2LHR': No such file or directory @ error/constitute.c/ReadImage/786.
```

**Reply by BeatVids** [reply]
Still thinking about what this means 🤔

**Comment by snibgo** [comment]
> magick a: r.png

That won't work because `a:` is not a built-in image.

**Reply by BeatVids** [reply]
Fixed, same error

**Comment by BeatVids** [comment]
magick rose: /data/data/com.termux/files/usr/tmp/magick-4GXcHbsfahGflSboUph-0SuocxEB2LHR.png

This works, so I do have access to the directory

**Comment by snibgo** [comment]
Okay, so it hasn't listed files, as the simple `ls` command does? Then IM has not successfully run the `ls` command. The problem isn't in Inkscape.

What is the text output from `magick -list policy`?

**Reply by BeatVids** [reply]
Looks like it is being listed I believe:

```
magick myls:r.png NULL:
r.png
magick: unable to open image '/data/data/com.termux/files/usr/tmp/magick-97ckgIHvOI9grKTV-KxmByDItDPGFmiG': No such file or directory @ error/blob.c/OpenBlob/3571.
magick: unable to open file '/data/data/com.termux/files/usr/tmp/magick-97ckgIHvOI9grKTV-KxmByDItDPGFmiG': No such file or directory @ error/constitute.c/ReadImage/786.
```

**Reply by BeatVids** [reply]
```
magick -list policy

Path: /data/data/com.termux/files/usr/etc/ImageMagick-7/policy.xml
  Policy: Undefined
    rights: None

Path: [built-in]
  Policy: Undefined
    rights: None
```

I'm guessing the `rights: None` is the culprit huh?

**Comment by snibgo** [comment]
> Looks like it is being listed I believe:

Ah, yes, if the only file in your current directory is r.png, then `ls` is working.

> I'm guessing the rights: None is the culprit huh?

No, that is a permissive policy. There is no problem there.

Sorry, I'm out of ideas. IM can run external delegates, but fails to run Inkscape, even though Inkscape can be run from the command line.

**Reply by BeatVids** [reply]
Thanks a lot to you and @fmw42 (I recognize him from like 10 years ago! What a fun tool ImageMagickis eh?) for trying to help me and my specific issue. I'll post a solution if I ever find anything

**Comment by fmw42** [comment]
Just to be sure, lets see what is actually in your policy

`cat /data/data/com.termux/files/usr/etc/ImageMagick-7/policy.xml`

Also show us your SVG delegate line in the delegates.xml file.  Does it reference Inkscape or RSVG or MSVG/XML

**Reply by BeatVids** [reply]
```
cat /data/data/com.termux/files/usr/etc/ImageMagick-7/policy.xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE policymap [
<!ELEMENT policymap (policy)*>
<!ATTLIST policymap xmlns CDATA #FIXED "">
<!ELEMENT policy EMPTY>
<!ATTLIST policy xmlns CDATA #FIXED "">
<!ATTLIST policy domain NMTOKEN #REQUIRED>
<!ATTLIST policy name NMTOKEN #IMPLIED>
<!ATTLIST policy pattern CDATA #IMPLIED>
<!ATTLIST policy rights NMTOKEN #IMPLIED>
<!ATTLIST policy stealth NMTOKEN #IMPLIED>
<!ATTLIST policy value CDATA #IMPLIED>
]>
<!--
  Creating a security policy that fits your specific local environment
  before making use of ImageMagick is highly advised. You can find guidance on
  setting up this policy at https://imagemagick.org/script/security-policy.php,
  and it's important to verify your policy using the validation tool located
  at https://imagemagick-secevaluator.doyensec.com/.


  Open ImageMagick security policy:

  The default policy for ImageMagick installations is the open security
  policy. This policy is designed for usage in secure settings like those
  protected by firewalls or within Docker containers. Within this framework,
  ImageMagick enjoys broad access to resources and functionalities. This policy
  provides convenient and adaptable options for image manipulation. However,
  it's important to note that it might present security vulnerabilities in
  less regulated conditions. Thus, organizations should thoroughly assess
  the appropriateness of the open policy according to their particular use
  case and security prerequisites.

  ImageMagick security policies in a nutshell:

  Domains include system, delegate, coder, filter, module, path, or resource.

  Rights include none, read, write, execute and all.  Use | to combine them,
  for example: "read | write" to permit read from, or write to, a path.

  Use a glob expression as a pattern.

  Suppose we do not want users to process MPEG video images, use this policy:

    <policy domain="delegate" rights="none" pattern="mpeg:decode" />

  Here we do not want users reading images from HTTP:

    <policy domain="coder" rights="none" pattern="HTTP" />

  The /repository file system is restricted to read only.  We use a glob
  expression to match all paths that start with /repository:

    <policy domain="path" rights="read" pattern="/repository/*" />

  Prevent users from executing any image filters:

    <policy domain="filter" rights="none" pattern="*" />

  Cache large images to disk rather than memory:

    <policy domain="resource" name="area" value="1GP"/>

  Use the default system font unless overridden by the application:

    <policy domain="system" name="font" value="/usr/share/fonts/favorite.ttf"/>

  Define arguments for the memory, map, area, width, height and disk resources
  with SI prefixes (.e.g 100MB).  In addition, resource policies are maximums
  for each instance of ImageMagick (e.g. policy memory limit 1GB, -limit 2GB
  exceeds policy maximum so memory limit is 1GB).

  Rules are processed in order.  Here we want to restrict ImageMagick to only
  read or write a small subset of proven web-safe image types:

    <policy domain="delegate" rights="none" pattern="*" />
    <policy domain="filter" rights="none" pattern="*" />
    <policy domain="coder" rights="none" pattern="*" />
    <policy domain="coder" rights="read|write" pattern="{GIF,JPEG,PNG,WEBP}" />

  See https://imagemagick.org/script/security-policy.php for a deeper
  understanding of ImageMagick security policies.
-->
<policymap>
  <policy domain="Undefined" rights="none"/>
  <!-- Set maximum parallel threads. -->
  <!-- <policy domain="resource" name="thread" value="2"/> -->
  <!-- Set maximum time to live in seconds or neumonics, e.g. "2 minutes". When
       this limit is exceeded, an exception is thrown and processing stops. -->
  <!-- <policy domain="resource" name="time" value="120"/> -->
  <!-- Set maximum number of open pixel cache files. When this limit is
       exceeded, any subsequent pixels cached to disk are closed and reopened
       on demand. -->
  <!-- <policy domain="resource" name="file" value="768"/> -->
  <!-- Set maximum amount of memory in bytes to allocate for the pixel cache
       from the heap. When this limit is exceeded, the image pixels are cached
       to memory-mapped disk. -->
  <!-- <policy domain="resource" name="memory" value="256MiB"/> -->
  <!-- Set maximum amount of memory map in bytes to allocate for the pixel
       cache. When this limit is exceeded, the image pixels are cached to
       disk. -->
  <!-- <policy domain="resource" name="map" value="512MiB"/> -->
  <!-- Set the maximum width * height of an image that can reside in the pixel
       cache memory. Images that exceed the area limit are cached to disk. -->
  <!-- <policy domain="resource" name="area" value="16KP"/> -->
  <!-- Set maximum amount of disk space in bytes permitted for use by the pixel
       cache. When this limit is exceeded, the pixel cache is not be created
       and an exception is thrown. -->
  <!-- <policy domain="resource" name="disk" value="1GiB"/> -->
  <!-- Set the maximum length of an image sequence.  When this limit is
       exceeded, an exception is thrown. -->
  <!-- <policy domain="resource" name="list-length" value="32"/> -->
  <!-- Set the maximum width of an image.  When this limit is exceeded, an
       exception is thrown. -->
  <!-- <policy domain="resource" name="width" value="8KP"/> -->
  <!-- Set the maximum height of an image.  When this limit is exceeded, an
       exception is thrown. -->
  <!-- <policy domain="resource" name="height" value="8KP"/> -->
  <!-- Periodically yield the CPU for at least the time specified in
       milliseconds. -->
  <!-- <policy domain="resource" name="throttle" value="2"/> -->
  <!-- Do not create temporary files in the default shared directories, instead
       specify a private area to store only ImageMagick temporary files. -->
  <!-- <policy domain="resource" name="temporary-path" value="/magick/tmp/"/> -->
  <!-- Force memory initialization by memory mapping select memory
       allocations. -->
  <!-- <policy domain="cache" name="memory-map" value="anonymous"/> -->
  <!-- Ensure all image data is fully flushed and synchronized to disk. -->
  <!-- <policy domain="cache" name="synchronize" value="true"/> -->
  <!-- Replace passphrase for secure distributed processing -->
  <!-- <policy domain="cache" name="shared-secret" value="secret-passphrase" stealth="true"/> -->
  <!-- Do not permit any delegates to execute. -->
  <!-- <policy domain="delegate" rights="none" pattern="*"/> -->
  <!-- Do not permit any image filters to load. -->
  <!-- <policy domain="filter" rights="none" pattern="*"/> -->
  <!-- Don't read/write from/to stdin/stdout. -->
  <!-- <policy domain="path" rights="none" pattern="-"/> -->
  <!-- don't read sensitive paths. -->
  <!-- <policy domain="path" rights="none" pattern="/etc/*"/> -->
  <!-- Indirect reads are not permitted. -->
  <!-- <policy domain="path" rights="none" pattern="@*"/> -->
  <!-- These image types are security risks on read, but write is fine -->
  <!-- <policy domain="module" rights="write" pattern="{MSL,MVG,PS,SVG,URL,XPS}"/> -->
  <!-- This policy sets the number of times to replace content of certain
       memory buffers and temporary files before they are freed or deleted. -->
  <!-- <policy domain="system" name="shred" value="1"/> -->
  <!-- Enable the initialization of buffers with zeros, resulting in a minor
       performance penalty but with improved security. -->
  <!-- <policy domain="system" name="memory-map" value="anonymous"/> -->
  <!-- Set the maximum amount of memory in bytes that are permitted for
       allocation requests. -->
  <!-- <policy domain="system" name="max-memory-request" value="256MiB"/> -->
</policymap>
```

**Reply by BeatVids** [reply]
```
grep 'svg' /data/data/com.termux/files/usr/etc/ImageMagick-7/delegates.xml
  <delegate decode="dot" command="&quot;dot&quot; -Tsvg &quot;%i&quot; -o &quot;%o&quot;" />
  <delegate decode="pnm" encode="trace" command="&quot;potrace&quot; --svg --output &quot;%o&quot; &quot;%i&quot;"/>
  <delegate decode="svg" command="&quot;rsvg-convert&quot; -o &quot;%o&quot; &quot;%i&quot;"/>
  <delegate decode="svg:decode" stealth="True" command="&quot;/data/data/com.termux/files/home/bin/inkscape&quot; &quot;%s&quot; --export-filename=&quot;%s&quot; --export-dpi=&quot;%s&quot; --export-background=&quot;%s&quot; --export-background-opacity=&quot;%s&quot; &gt; &quot;%s&quot; 2&gt;&amp;1"/>
```

**Comment by fmw42** [comment]
Your SVG policy is commented. Perhaps uncomment it and add "read" to it. So change

`  <!-- <policy domain="module" rights="write" pattern="{MSL,MVG,PS,SVG,URL,XPS}"/> -->`

 to 

`<policy domain="module" rights="read|write" pattern="{MSL,MVG,PS,SVG,URL,XPS}"/> `

**Reply by BeatVids** [reply]
```
grep SVG /data/data/com.termux/files/usr/etc/ImageMagick-7/policy.xml
<policy domain="module" rights="read|write" pattern="{MSL,MVG,PS,SVG,URL,XPS}"/>
```

```
magick  in.svg out.png
magick: FailedToExecuteCommand `'/data/data/com.termux/files/usr/bin/inkscape' '/data/data/com.termux/files/usr/tmp/magick-fs1ri7Wg1gA3lLZVB8h16440ItIbc2B5' --export-filename='/data/data/com.termux/files/usr/tmp/magick-PxystgVz_OyVsSU712Wbet0guztzkhEn.png' --export-dpi='96' --export-background='rgb(100%,100%,100%)' --export-background-opacity='1' > '/data/data/com.termux/files/usr/tmp/magick-5NbXAw_lY2BJMMcTv16B8D5baOfrpbkC' 2>&1' (-1) @ error/delegate.c/ExternalDelegateCommand/522.
```

**Reply by BeatVids** [reply]
Dang I had my hopes up with this one!

**Reply by fmw42** [reply]
Are you sure you have the correct location for Inkscape? Do you have more than one policy.xml?

**Reply by BeatVids** [reply]
I have `/data/data/com.termux/files/usr/share/doc/ImageMagick-7/www/source/policy.xml`, but there's no mention of SVG anywhere in the file, it's different than `/data/data/com.termux/files/usr/etc/ImageMagick-7/policy.xml`

Yes, I'm using the correct location
```
whereis inkscape
inkscape: /data/data/com.termux/files/usr/bin/inkscape
```

**Reply by fmw42** [reply]
Actual policy would be in /etc, but could be elsewhere. See https://imagemagick.org/script/resources.php

 The /share is not what you want.  But do you have more than one location for Inkscape?

**Reply by BeatVids** [reply]
Yes, that is the only location. The man page came up on the whereis, but I ommitted it to simplify.

I don't have an x server running, and I think inkscape uses it, but still outputs files no problem, but it might be a consideration?

**Reply by fmw42** [reply]
Where is Imagemagick "magick" installed?

**Reply by BeatVids** [reply]
```
whereis magick
magick: /data/data/com.termux/files/usr/bin/magick
```

**Comment by BeatVids** [comment]
@fmw42 @snibgo 

Updated the delegates.xml to a simplified command:
`<delegate decode="svg:decode" command="&quot;/data/data/com.termux/files/usr/bin/inkscape&quot;"/>`

And `magick in.svg out.png` produces out.png! Of course, missing the arguments, but I'm going to find out the main culprit.

I appreciate you guys taking the time to help and your valuable perspectives!

**Comment by BeatVids** [comment, marked as answer]
Removing the very last 2 parts did it:
`<delegate decode="svg:decode" stealth="True" command="&quot;/data/data/com.termux/files/usr/bin/inkscape&quot; &quot;%s&quot; --export-filename=&quot;%s&quot; --export-dpi=&quot;%s&quot; --export-background=&quot;%s&quot; --export-background-opacity=&quot;%s&quot; &gt; &quot;%s&quot; 2&gt;&amp;1"/>
`

To this:
`<delegate decode="svg:decode" stealth="True" command="&quot;/data/data/com.termux/files/usr/bin/inkscape&quot; &quot;%s&quot; --export-filename=&quot;%s&quot; --export-dpi=&quot;%s&quot; --export-background=&quot;%s&quot; --export-background-opacity=&quot;%s&quot;"/>
`

That part of the command looks like this:
> '/data/data/com.termux/files/usr/tmp/magick-2NjJmBk7VGtoFx4rmnjyZ8PgtIWZZiub' 2>&1

I know the last part suppresses errors, but I'm not sure what the consequences will be from ommitting the first part that outputs to file

**Comment by snibgo** [comment]
Thanks for your persistence. The fix is interesting.

`2>&1` redirects stderr to stdout, so text error messages will go to stdout. Just before that, `> filename` redirects stdout to the named file.

The redirection is handled by the shell, not by IM or Inkscape. I suppose there is something about Android that works differently to Unix and Windows.

As far as I know, IM does nothing with the file from stdout and stderr, so I expect removing this is harmless.

**Reply by BeatVids** [reply]
Probably also in combination that it's in a Termux environment in an unrooted device!
